### PR TITLE
Wire protocol 

### PR DESCRIPTION
**Pull request by moshababo**
Addressing #284, and also changes the way we handle multiple packets stream.

I preferred not to use anymore the `\r\n` sign anymore, because it can potentially be found in a string property as well. Instead, the packet length is specified.

Putting this inside of the packet JSON, in the header, was not possible (packet string must be complete in order to be parsable), so I added it as the first 4 bytes in the stream.

I couldn't reproduce yet the packet splitting - couldn't connect to the remote nodes (my ports are blocked), and locally it never happen. Will do it soon. 
Wanted to open this PR so you could give feedback on the approach.

If the approach is acceptable, I can add some length constrains for safety.

### CODE REVIEW

File: lib/p2p/Parser.ts [hunk 1]
```diff
@@ -60,20 +60,46 @@ interface Parser {
 
 /** Protocol packet parser */
 class Parser extends EventEmitter {
+  private pending: Buffer[] = [];
+  private waiting: number = 0;
```
Comment (sangaman) [inline lib/p2p/Parser.ts]: These names aren't that clear. I think maybe `pendingBuffers` or `pendingPacket` for the first and `remainingBytes` for the second are a couple of suggestions.

File: lib/p2p/Parser.ts [hunk 2]
```diff
@@ -60,20 +60,46 @@ interface Parser {
...
+      this.parsePacket(this.pending);
+      this.resetCycle();
+      if (size < chunk.length) { // multiple packets
+        this.feed(chunk.slice(size));
       }
-    });
+    }
+  }
+
+  private resetCycle = () => {
```
Comment (sangaman) [inline lib/p2p/Parser.ts]: Maybe call this `resetBuffer`? Or maybe better just put these two lines into the `read` function after we parse a packet, unless you think it's going to be called elsewhere later on.

File: lib/p2p/Parser.ts [hunk 3]
```diff
@@ -60,20 +60,46 @@ interface Parser {
...
+  private read = (size: number, chunk: Buffer) => {
+    this.pending.push(chunk.slice(0, size));
+
+    if (size > chunk.length) { // packet isn't complete
+      this.waiting = size - chunk.length;
+    } else { // chunk is finalizing the packet
+      this.parsePacket(this.pending);
+      this.resetCycle();
+      if (size < chunk.length) { // multiple packets
+        this.feed(chunk.slice(size));
```
Comment (offerm) [inline lib/p2p/Parser.ts]: are you taking here the first part or the left over?
also, what if the leftover size is <4? you will not be able to read the length.

**Comment by offerm** [comment]
See this.
https://stackoverflow.com/questions/17312242/node-js-net-library-getting-complete-data-from-data-event
I think the first answer shows correctly how this should be done

**Comment by moshababo** [comment]
I can handle the scenario where chunks would split on the 4 bytes. Also, I can add another 1 byte flag, following the 4 bytes, to ensure that i'm not reading just a random 4 bytes and interpreting it as a number (and if I am, the chunk would be dropped).

Or, we can use the delimiter approach. With specifying packet size, I know the size of the packet from the beginning, so I can drop it it if it's too big. With the delimiter approach, I have to reach my maximum before I know about it.
Also, the delimiter is a bit less safe - even if we'll use a unique one, it will be known, and can be injected somehow into one of the strings that a node might use in packets. 

So in the overall, specifying size is more complicated, but a bit more safe. Both can be used.
I'm used to seeing size specifying in p2p protocols. Not using it was an option only because we're using JSON serialization. 

So tell me what you think, and i'll finalize this.

**Comment by moshababo** [comment]
I can handle the scenario where chunks would split on the 4 bytes. Also, I can add another 1 byte flag, following the 4 bytes, to ensure that i'm not reading just a random 4 bytes and interpreting it as a number (and if I am, the chunk would be dropped).

Or, we can use the delimiter approach. With specifying packet size, I know the size of the packet from the beginning, so I can drop it it if it's too big. With the delimiter approach, I have to reach my maximum before I know about it.
Also, the delimiter is a bit less safe - even if we'll use a unique one, it will be known, and can be injected somehow into one of the strings that a node might use in packets.

So in the overall, specifying size is more complicated, but a bit more safe. Both can be used.
I'm used to specifying size in p2p protocols. Not using it was an option only because we're using JSON serialization.

So tell me what you think, and i'll finalize this.

**Comment by offerm** [comment]
as you wrote both approaches should work.

Looks to me that `size` works better when you can read from the connection directly and you can specify the amount you want to read. This way, once you know the size of the message, you can be sure you are not reading anything belonging to the next message. Also, there is normally an `available()` function that you can use to be sure that there are at least 4 bytes ready to be read.

In our case, the `data` is a callback. In this case, I think a separator will work better.

you should do something like this:
```
    conn.on('data', function(data) {
        completeData += data;
        var dataArray = completeData.split('your delimiter');
        if(dataArray.size > 1) { //If our data was split into several pieces, we have a complete chunk saved in the 0th position in the array
            doWorkOnTheFirstHalfOfData(dataArray[0]);
            completeData = dataArray[1];// The second portion of data may yet be incomplete, thise may need to be more complete logic if you can get more than one delimeter at a time...
        }
    });
```
Basically, you append the received data. If there is a separator in the data you handle all chunks to the left of the last separator. Note that the code above is not doing that properly and is not ready for getting multiple separators in one data part.

**Comment by moshababo** [comment]
There's no problem of reading min/max length, because the buffer array length is known. So I didn't understand the argument against it.

**Comment by offerm** [comment]
Try to write the code and you will see.
What if you get a 3 bytes buffer on the first feed() call?

**Comment by moshababo** [comment]
Then I wait for the next chunk, and only then try to parse the number. Before I have 4 bytes, i'm iterating until i'll get them. After I have it, i'm iterating until i'll consume the packet.
It will add some more code, and I didn't write it yet, but I don't see a problem with the approach.

**Comment by offerm** [comment]
As I wrote above

> as you wrote both approaches should work.

Do it as you like. My feeling is that the seperator option will lead a simple code which is way easier to understand.

**Comment by moshababo** [comment]
Code simplicity is important, but the problem we were discussing couldn't even be solved if we'll use a separator.

**Comment by moshababo** [comment]
me and @offerm talked, and I was wrong, the problem can be addressed while using a separator, while having an overall simpler solution. 

@sangaman waiting for your take on this to finalize.

**Comment by moshababo** [comment]
Yes, i'll do the necessary refactoring.

**Comment by moshababo** [comment]
Latest commit:
* changed the parser to use delimiter. it's fairly simple now, only 7 lines
* added some tests
* refactored `Packet` a bit. mostly due to the deep equal checks in the tests

File: lib/p2p/Parser.ts [hunk 4]
```diff
@@ -60,20 +60,30 @@ interface Parser {
...
-        return;
-      }
-      try {
-        const packet = fromRaw(packetStr);
-        this.emit('packet', packet);
-      } catch (err) {
-        this.emit('error', err);
-      }
-    });
+    this.pending += data.toString();
```
Comment (offerm) [inline lib/p2p/Parser.ts]: Suggest to protect the max size of this.pending to avoid accumulation of a very large message

File: lib/p2p/Parser.ts [hunk 5]
```diff
@@ -60,20 +60,30 @@ interface Parser {
...
+    const index = this.pending.indexOf(this.delimiter);
+    if (index > -1) {
```
Comment (offerm) [inline lib/p2p/Parser.ts]: Suggest to have tests for:
- empty Buffer
- buffer containing one delimiter only (no other data)
- buffer containing 5  delimiters (no other data)

File: lib/p2p/Parser.ts (hunk 4 above)
Comment (sangaman) [reply lib/p2p/Parser.ts]: I was thinking of having this implemented later but maybe not too hard to take care of it here, what's a reasonable number? 4 MB? I can imagine the ORDERS packet getting fairly large.

File: lib/p2p/Parser.ts [hunk 6]
```diff
@@ -60,20 +60,30 @@ interface Parser {
 
 /** Protocol packet parser */
 class Parser extends EventEmitter {
+  private pending = '';
```
Comment (sangaman) [inline lib/p2p/Parser.ts]: I raised a similar comment earlier but I'd prefer calling this `buffer` or maybe `pendingPacket`. Actually, since it might contain multiple packets I don't think `pendingPacket` would be too good, but I still like `buffer`.

File: lib/p2p/packets/Packet.ts [hunk 7]
```diff
@@ -40,6 +40,8 @@ abstract class Packet<T = any> implements PacketInterface {
   public abstract get direction(): PacketDirection;
   public body?: T;
   public header: PacketHeader;
+  /** Wire protocol delimiter. */
+  public static PROTOCOL_DELIMITER = 'ↂ₿ↂ';
```
Comment (sangaman) [inline lib/p2p/packets/Packet.ts]: Nice! What are those 2 things on the side?

File: lib/p2p/Parser.ts [hunk 8]
```diff
@@ -60,20 +60,30 @@ interface Parser {
...
-      }
-      try {
-        const packet = fromRaw(packetStr);
-        this.emit('packet', packet);
-      } catch (err) {
-        this.emit('error', err);
-      }
-    });
+    this.pending += data.toString();
+    const index = this.pending.indexOf(this.delimiter);
```
Comment (sangaman) [inline lib/p2p/Parser.ts]: I have a minor performance concern here where `indexOf` will get slower and slower as `pending` gets larger, but I realize it's not as simple as checking the `data.toString()` for the delimiter because the delimiter could be split between two chunks. I'm pretty sure this is O(n^2) complexity with the size of the packet and something we'll be doing very often so I'm just concerned and maybe we should at least make a note to revisit this.

One solution that just came to mind is tracking the last received chunk and the current chunk, and concatenating those then checking the result for the delimiter. We'd clear the last chunk on creating a packet.

Not necessary for this PR, but something to think about.

File: lib/p2p/Parser.ts [hunk 9]
```diff
@@ -60,20 +60,30 @@ interface Parser {
...
+    if (index > -1) {
+      this.parsePacket(this.pending.slice(0, index));
+      const next = this.pending.slice(index + this.delimiter.length);
+      this.pending = '';
+      this.feed(Buffer.from(next));
```
Comment (sangaman) [inline lib/p2p/Parser.ts]: We're converting a buffer to a string and then back to a buffer here, I'm thinking it might be better to make `feed()` accept a string parameter and do `toString()` from where we call the method in Peer.ts. It would also make the tests a little cleaner, since we have `Buffer.from` calls in multiple places that we wouldn't need if `feed` took a string.

File: test/unit/Parser.spec.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  testPacket(pingPacket);
+  testPacket(helloPacket);
+
+  testSplitPacket(pingPacket);
+  testSplitPacket(helloPacket);
+
+  testConcatenatedPackets([pingPacket, helloPacket, pingPacket]);
+
+  testConcatenatedAndSplitOnTheDelimiter([pingPacket, helloPacket, pingPacket]);
+
```
Comment (sangaman) [inline test/unit/Parser.spec.ts]: Can drop this blank line.

File: lib/p2p/packets/Packet.ts [hunk 11]
```diff
@@ -56,15 +58,20 @@ abstract class Packet<T = any> implements PacketInterface {
   constructor(bodyOrPacket?: T | PacketInterface, reqId?: string) {
     if (isPacketInterface(bodyOrPacket)) {
       // we are deserializing a received packet from a raw JSON string
-      this.body = bodyOrPacket.body;
+      delete bodyOrPacket.header.type; // deleting type from header because it's being used only for the wire protocol
```
Comment (sangaman) [inline lib/p2p/packets/Packet.ts]: This is smart.

File: lib/p2p/packets/Packet.ts [hunk 12]
```diff
@@ -56,15 +58,20 @@ abstract class Packet<T = any> implements PacketInterface {
...
       this.header = bodyOrPacket.header;
+
+      if (bodyOrPacket.body) {
+        this.body = bodyOrPacket.body;
```
Comment (sangaman) [inline lib/p2p/packets/Packet.ts]: What's the advantage of this over `this.body = bodyOrPacket.body`? I think it winds up undefined either way when the body is undefined.

Comment (moshababo) [reply lib/p2p/packets/Packet.ts]: if it's undefined, then the property is set to undefined. it's not the same as if the property wouldn't be set at all. So if doing a deep equal check on the objects, it fails. It's possible to just remove the undefined properties before doing the equal check, but I thought that this would be better.

File: lib/p2p/Parser.ts (hunk 9 above)
Comment (moshababo) [reply lib/p2p/Parser.ts]: I left it as is because I thought that maybe at some point we'll want to read the bytes array and not the string. But I can do what you suggester for now.

File: lib/p2p/packets/Packet.ts (hunk 7 above)
Comment (moshababo) [reply lib/p2p/packets/Packet.ts]: Just some random thing I found on the unicode table. I think it's a roman number. Feel free to change the delimiter if you want.

File: lib/p2p/Parser.ts (hunk 6 above)
Comment (moshababo) [reply lib/p2p/Parser.ts]: I didn't want `buffer` to be confused with the `Buffer` type. But I can use it if you prefer, since `feed` will receive a string anyway.

Comment (moshababo) [reply lib/p2p/Parser.ts]: I didn't want `buffer` to be confused with the `Buffer` type. But I can use it if you prefer, since `feed` will receive a string now.

File: lib/p2p/Parser.ts (hunk 8 above)
Comment (moshababo) [reply lib/p2p/Parser.ts]: Your solution could improve it, you can open a separate issue for that if you want. Not high priority at the moment.

**Comment by moshababo** [comment]
Latest commits:
* parser now receives a string
* renamed `pending` to `buffer`
* max buffer size support added, default to 4 MB. we can open an issue for making it configurable, if we want
* test code improved, and more test were added

The travis build results are not consistent. It's sometime fails on the max buffer size checks.
JS string supposed to be UTF-16, 2 bytes. But I checked it with `Buffer.byteLength`, and it was only 1 byte. I ended up using it for the actual max length checks. 
If it still doing problems, it means that converting a fixed size buffer to string, and then calculating the string byte length, results in a different size. 
If that's the case, I'll need to find a different solution. I just want to make sure it's not just another travis issue.

File: lib/p2p/packets/Packet.ts (hunk 12 above)
Comment (sangaman) [reply lib/p2p/packets/Packet.ts]: Ah yeah that's what you meant by deep checks, ok cool, thanks.

File: lib/p2p/packets/Packet.ts (hunk 7 above)
Comment (sangaman) [reply lib/p2p/packets/Packet.ts]: Nah it's fine, as long as it works.

**Comment by sangaman** [comment]
The default encoding when you do Buffer.toString() is utf-8. Same for Buffer.from(), it assumes utf-8 unless specified otherwise. In some quick tests I did, the `ↂ₿ↂ` comes out as 9 bytes with utf-8 and `abc` comes out as 3 bytes. It does look like you can't easily calculate the bytes of a string - you have to convert back to buffer first I think. You could maybe keep a running number tally of the number of bytes along with the string buffer perhaps. Ultimately I don't think this needs to hold up this PR unless it's consistently giving us build issues - do the tests fail occasionally for you when running locally? I'll try it myself later.

Also 9 bytes sounds like kind of a lot for a separator, but it's not a big deal.

File: test/unit/Parser.spec.ts [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    parser.feed(delimiter);
+  });
+
+  it(`should buffer a max buffer length`, (done) => {
+    parser = new Parser(delimiter, 10);
+    wait()
+      .then(() => expect.fail())
+      .catch((err) => {
+        expect(err).to.equal('timeout');
```
Comment (sangaman) [inline test/unit/Parser.spec.ts]: I'm getting an unhandled rejection error from this line not every time I run the tests locally but most times. If you merge master, I'm using the `chai-as-promised` plugin for some of the sanity tests. Maybe you can use it here as well with the sanity tests as an example.

**Comment by sangaman** [comment]
So I actually did a rebase myself and refactored two of the test cases and now the tests pass every time for me, but I'm still seeing some unhandled rejection warnings that I assume are coming from other test cases. Here's what the refactored test cases look like:

```javascript
  it(`should try parse a standalone delimiter and fail`, () => {
    expect(wait()).to.be.rejectedWith(ParserErrorType.UNPARSEABLE_MESSAGE.toString());

    parser.feed(delimiter);
  });

  it(`should buffer a max buffer length`, () => {
    parser = new Parser(delimiter, 10);
    expect(wait()).to.eventually.be.fulfilled;

    parser.feed(Buffer.allocUnsafe(10).toString());
  });
```

If you want I can push my changes to the branch and you can finish it, let me know.

File: test/unit/Parser.spec.ts (hunk 13 above)
Comment (moshababo) [reply test/unit/Parser.spec.ts]: The 'timeout' exception is expected. If you're talking about tests timeout (20 sec), then it doesn't happen for me locally.
I just saw your `chai-as-promised` usage - looks nice.

**Comment by sangaman** [comment]
OK cool, well the tests pass for me consistently now. If we can expose the parser error objects and maybe clean up the unhandled exception warnings in the test output in a future PR, that would be great, but I think this is good for now. Thanks.
